core/src/server/subscription.rs

```diff
@@ -0,0 +1,475 @@
+	}
+
+	pub(crate) fn from_complete_message(msg: String) -> Self {
+		SubscriptionMessage(SubscriptionMessageInner::Complete(msg))
+	}
+
+	pub(crate) fn empty() -> Self {
+		Self::from_complete_message(String::new())
+	}
+}
+
+impl<T> From<T> for SubscriptionMessage
+where
+	T: AsRef<str>,
+{
+	fn from(msg: T) -> Self {
+		SubscriptionMessage(SubscriptionMessageInner::NeedsData(format!("\"{}\"", msg.as_ref())))
+	}
+}
+
+/// Represent a unique subscription entry based on [`SubscriptionId`] and [`ConnectionId`].
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct SubscriptionKey {
+	pub(crate) conn_id: ConnectionId,
+	pub(crate) sub_id: SubscriptionId<'static>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum SubNotifResultOrError {
+	Result,
+	Error,
+}
+
+impl SubNotifResultOrError {
+	pub(crate) const fn as_str(&self) -> &str {
+		match self {
+			Self::Result => "result",
+			Self::Error => "error",
+		}
+	}
+}
+
+/// Represents a subscription until it is unsubscribed.
+///
+// NOTE: The reason why we use `mpsc` here is because it allows `IsUnsubscribed::unsubscribed`
+// to be &self instead of &mut self.
+#[derive(Debug, Clone)]
+pub struct IsUnsubscribed(mpsc::Sender<()>);
+
+impl IsUnsubscribed {
+	/// Returns true if the unsubscribe method has been invoked or the subscription has been canceled.
+	///
+	/// This can be called multiple times as the element in the channel is never
+	/// removed.
+	pub fn is_unsubscribed(&self) -> bool {
+		self.0.is_closed()
+	}
+
+	/// Wrapper over [`tokio::sync::mpsc::Sender::closed`]
+	///
+	/// # Cancel safety
+	///
+	/// This method is cancel safe. Once the channel is closed,
+	/// it stays closed forever and all future calls to closed will return immediately.
+	pub async fn unsubscribed(&self) {
+		self.0.closed().await;
+	}
+}
+
+/// Represents a single subscription that is waiting to be accepted or rejected.
+///
+/// If this is dropped without calling `PendingSubscription::reject` or `PendingSubscriptionSink::accept`
+/// a default error is sent out as response to the subscription call.
+///
+/// Thus, if you want a customized error message then `PendingSubscription::reject` must be called.
+#[derive(Debug)]
+#[must_use = "PendningSubscriptionSink does nothing unless `accept` or `reject` is called"]
+pub struct PendingSubscriptionSink {
+	/// Sink.
+	pub(crate) inner: MethodSink,
+	/// MethodCallback.
+	pub(crate) method: &'static str,
+	/// Shared Mutex of subscriptions for this method.
+	pub(crate) subscribers: Subscribers,
+	/// Unique subscription.
+	pub(crate) uniq_sub: SubscriptionKey,
+	/// ID of the `subscription call` (i.e. not the same as subscription id) which is used
+	/// to reply to subscription method call and must only be used once.
+	pub(crate) id: Id<'static>,
+	/// Sender to answer the subscribe call.
+	pub(crate) subscribe: mpsc::Sender<MethodResponse>,
+	/// Subscription permit.
+	pub(crate) permit: SubscriptionPermit,
```

Comment:
Nit: I'm wondering how many of these need to be `pub(crate)` and whether we can hide a bit more

Reply:
it's instantiated from the `RpcModule` so needs to be `pub(crate)` or having `pub(crate) fn new`